pkg/index/index.go

```diff
@@ -0,0 +1,172 @@
+	Name       string   `yaml:"name"`
+	Registry   string   `yaml:"registry"`
+	Repository string   `yaml:"repository"`
+	Sources    []string `yaml:"sources"`
+	Type       string   `yaml:"type"`
+}
+
+// Index represents an index.
+type Index struct {
+	Name        string
+	Filename    string
+	Entries     []*Entry
+	entryByName map[string]*Entry
+}
+
+// MergedIndexes is used to aggregate all indexes and perform search operations.
+type MergedIndexes struct {
+	Index
+	indexByEntry map[*Entry]*Index
+}
+
+// NewIndex loads an Index from a file.
+func NewIndex(path, name string) (*Index, error) {
+	indexBytes, err := os.ReadFile(filepath.Clean(path))
+	if err != nil {
+		return nil, fmt.Errorf("cannot read file %s: %w", path, err)
+	}
+
+	var index Index
+	var entries []Entry
+	if err = yaml.Unmarshal(indexBytes, &entries); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal index: %w", err)
+	}
+
+	for k := range entries {
+		index.Entries = append(index.Entries, &entries[k])
+	}
+
+	index.Name = name
+	index.Filename = path
+
+	return &index, nil
+}
+
+// Upsert adds a new entry to the Index or updates an existing one.
+func (i *Index) Upsert(entry *Entry) {
+	defer func() {
+		i.entryByName[entry.Name] = entry
+	}()
+
+	for k, e := range i.Entries {
+		if e.Name == entry.Name {
+			i.Entries[k] = entry
+			return
+		}
+	}
+
+	i.Entries = append(i.Entries, entry)
+}
+
+// Remove removes an entry from the Index.
+func (i *Index) Remove(entry *Entry) error {
+	for k, e := range i.Entries {
+		if e == entry {
+			i.Entries = append(i.Entries[:k], i.Entries[k+1:]...)
+			delete(i.entryByName, e.Name)
+			return nil
+		}
+	}
+
+	return fmt.Errorf("cannot remove %s: not found", entry.Name)
+}
+
+// EntryByName returns a Entry by passing its name.
+func (i *Index) EntryByName(name string) *Entry {
+	return i.entryByName[name]
+}
+
+// Write writes an Index to disk.
+func (i *Index) Write(path string) error {
+	indexBytes, err := yaml.Marshal(i.Entries)
+	if err != nil {
+		return fmt.Errorf("cannot marshal index: %w", err)
+	}
+
+	if err = os.WriteFile(path, indexBytes, writePermissions); err != nil {
+		return fmt.Errorf("cannot write index to file: %w", err)
+	}
+
+	return nil
+}
+
+// NewMergedIndexes initializes a MergedIndex.
+func NewMergedIndexes() *MergedIndexes {
+	m := &MergedIndexes{}
+
+	m.entryByName = make(map[string]*Entry)
+	m.indexByEntry = make(map[*Entry]*Index)
+
+	return m
+}
+
+// Merge creates a new index by merging all the indexes that are passed.
+// Orders matters. Be sure to pass an ordered list of indexes. For our use case, sort by added time.
+func (m *MergedIndexes) Merge(indexes ...*Index) {
+	for _, index := range indexes {
+		for _, Entry := range index.Entries {
+			m.Upsert(Entry)
+			m.indexByEntry[Entry] = index
+		}
+	}
+}
+
+// SearchByKeywords search for entries matching the given keywords in MergedIndexes.
+func (m *MergedIndexes) SearchByKeywords(keywords ...string) []*Entry {
+	var result []*Entry
```

Comment:
We may move this up to `Index`, and `MergedIndexes` inherits.

Reply:
I like this, thanks!